spark/src/test/scala/org/apache/spark/sql/delta/DeltaProtocolVersionSuite.scala

```diff
@@ -588,6 +588,167 @@ trait DeltaProtocolVersionSuiteBase extends QueryTest
+        TABLE_FEATURES_MIN_WRITER_VERSION)
+        .withWriterFeatures(Seq("NonExistingWriterFeature1", "NonExistingWriterFeature2"))
+      untrackedChangeProtocolVersion(deltaLog, version, protocolWriterFeatures)
+
+      val exceptionWrite = intercept[DeltaUnsupportedTableFeatureException] {
+        spark.range(1).write
+          .mode("append")
+          .option("mergeSchema", "true")
+          .format("delta")
+          .save(path.getCanonicalPath)
+      }
+      assert(exceptionWrite.getMessage ==
+        getExpectedWriterFeatureErrorMessage(
+          deltaLog.dataPath.toString,
+          "NonExistingWriterFeature1, NonExistingWriterFeature2"))
+    }
+  }
+
+  def testTableFeatureErrorMessageWithTableName(warm: Boolean): Unit = {
+    val featureTable = "mytablefeaturesnotsupported"
+    withTable(featureTable) {
+      spark.range(1).write.format("delta").saveAsTable(featureTable)
+      val (deltaLog, snapshot) = DeltaLog.forTableWithSnapshot(spark, TableIdentifier(featureTable))
+
+      var version = snapshot.version + 1
+      val protocolReaderFeatures = Protocol(
+        TABLE_FEATURES_MIN_READER_VERSION,
+        TABLE_FEATURES_MIN_WRITER_VERSION)
+        .withReaderFeatures(Seq("NonExistingReaderFeature1", "NonExistingReaderFeature2"))
+      untrackedChangeProtocolVersion(deltaLog, version, protocolReaderFeatures)
+      if (!warm) {
+        DeltaLog.clearCache()
+      }
+
+      val exceptionRead = intercept[DeltaUnsupportedTableFeatureException] {
+        spark.read.format("delta").table(featureTable)
+      }
+      val pathInErrorMessage = "default." + featureTable
+      assert(exceptionRead.getMessage ==
+        getExpectedReaderFeatureErrorMessage(
+          pathInErrorMessage,
+          "NonExistingReaderFeature1, NonExistingReaderFeature2"))
+
+      version = version + 1
+      val protocolWriterFeatures = Protocol(
+        TABLE_FEATURES_MIN_READER_VERSION,
+        TABLE_FEATURES_MIN_WRITER_VERSION)
+        .withWriterFeatures(Seq("NonExistingWriterFeature1", "NonExistingWriterFeature2"))
+      untrackedChangeProtocolVersion(deltaLog, version, protocolWriterFeatures)
+      if (!warm) {
+        DeltaLog.clearCache()
+      }
+
+      val exceptionWrite = intercept[DeltaUnsupportedTableFeatureException] {
+        spark.range(1).write
+          .mode("append")
+          .option("mergeSchema", "true")
+          .format("delta")
+          .saveAsTable(featureTable)
+      }
+      assert(exceptionWrite.getMessage ==
+        getExpectedWriterFeatureErrorMessage(
+          pathInErrorMessage,
+          "NonExistingWriterFeature1, NonExistingWriterFeature2"))
+
+      // Restore the protocol version or the clean-up fails
+      version = version + 1
+      untrackedChangeProtocolVersion(deltaLog, version, 1, 2)
+    }
+  }
+
+  test("DeltaUnsupportedTableFeatureException - error message with table name - warm") {
+    testTableFeatureErrorMessageWithTableName(true)
```

Comment:
Named argument, here and below? Otherwise it's hard to verify that "true" actually corresponds to "warm".